makefile

```diff
@@ -33,41 +40,78 @@ override DMD_FLAGS += $(DFLAGS)
 override LDC_FLAGS += $(DFLAGS)
 override GDC_FLAGS += $(DFLAGS)
 DMD_TEST_FLAGS = -w -g -Jbin -version=StdLoggerDisableWarning
-override LDC_FLAGS += -O5 -release -oq -d-version=StdLoggerDisableWarning
-override GDC_FLAGS += -O3 -frelease -d-version=StdLoggerDisableWarning
+override LDC_FLAGS += -O5 -release -oq -d-version=StdLoggerDisableWarning -Jbin
+override GDC_FLAGS += -O3 -frelease -d-version=StdLoggerDisableWarning -Jbin
 SHELL:=/usr/bin/env bash
 
-all: dmdbuild
-ldc: ldcbuild
-gdc: gdcbuild
+DMD_BIN = bin/dmd/dscanner
+LDC_BIN = bin/ldc/dscanner
+GDC_BIN = bin/gdc/dscanner
+
+GITHASH = bin/githash.txt
+
+OBJ_BY_DMD = $(addprefix $(OBJ_DIR)/dmd/, $(OBJ))
+UT_OBJ_BY_DMD = $(addprefix $(UT_OBJ_DIR)/dmd/, $(PROJECT_OBJ))
+
+OBJ_BY_LDC = $(addprefix $(OBJ_DIR)/ldc/, $(OBJ))
+UT_OBJ_BY_LDC = $(addprefix $(UT_OBJ_DIR)/ldc/, $(PROJECT_OBJ))
+
+OBJ_BY_GDC = $(addprefix $(OBJ_DIR)/gdc/, $(OBJ))
+UT_OBJ_BY_GDC = $(addprefix $(UT_OBJ_DIR)/gdc/, $(PROJECT_OBJ))
+
+$(OBJ_DIR)/dmd/%.o: %.d
+	@test -d $(dir $@) || mkdir -p $(dir $@)
```

Comment:
Use a [order-only dependency](https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html#Prerequisite-Types) for the directory